### ISSUE TITLE
Server Setup Documentation

### ISSUE THREAD

**Issue by jborean93** [jborean93/pypsrp#26]
_From @danielNemeth19 https://github.com/jborean93/pypsrp/issues/23#issuecomment-464228114_

Would you be able to point me to a doc that explains the recommended settings config wise that one needs to do on the windows machine for remote queries to run? I was following this guide https://4sysops.com/archives/powershell-remoting-between-windows-and-linux/ and also these ones: https://github.com/PowerShell/psl-omi-provider and https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/enable-psremoting?view=powershell-6. My major worry is that most of the manuals I've seen (there were few others) are saying that I need to set basic auth and unnencrypted connections to true, trusted hosts to *,  while also mentioning that by the way these settings are not secure / recommended at all. I've read about using self-signed SSL certificates over HTTPS (https://4sysops.com/archives/powershell-remoting-over-https-with-a-self-signed-ssl-certificate/) but the setup seems to be quite daunting and what if I don't really have control over the remote machines? I guess my question is how safe is to use to use basic auth?

**Comment by jborean93** [comment]
> Would you be able to point me to a doc that explains the recommended settings config wise that one needs to do on the windows machine for remote queries to run?

You don't/shouldn't need to do anything different from the basic Windows configuration to get pypsrp working. This library supports pretty much all the connection and authentication options, the only remaining things I know of is compression support. For PSDesktop just run [Enable-PSRemoting](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/enable-psremoting?view=powershell-6) and it will do the work for you. In fact on 2012 R2 and newer you shouldn't have to do anything as it will be enabled by default.

PSCore is a bit of a different story, I believe in the msi installer there's a checkbox to set this up the WSMan endpoint for you, if not I believe `Enable-PSRemoting` is also used for WSMan. When in doubt you can run the cmdlet `Get-PSSessionConfiguration`, this will show all the registered WSMan endpoints.

```
PS C:\Windows\system32> Get-PSSessionConfiguration


Name          : microsoft.powershell
PSVersion     : 5.1
StartupScript :
RunAsUser     :
Permission    : NT AUTHORITY\INTERACTIVE AccessAllowed, BUILTIN\Administrators AccessAllowed, BUILTIN\Remote
                Management Users AccessAllowed

Name          : microsoft.powershell.workflow
PSVersion     : 5.1
StartupScript :
RunAsUser     :
Permission    : BUILTIN\Administrators AccessAllowed, BUILTIN\Remote Management Users AccessAllowed

Name          : microsoft.powershell32
PSVersion     : 5.1
StartupScript :
RunAsUser     :
Permission    : NT AUTHORITY\INTERACTIVE AccessAllowed, BUILTIN\Administrators AccessAllowed, BUILTIN\Remote
                Management Users AccessAllowed

Name          : microsoft.windows.servermanagerworkflows
PSVersion     : 3.0
StartupScript :
RunAsUser     :
Permission    : NT AUTHORITY\INTERACTIVE AccessAllowed, BUILTIN\Administrators AccessAllowed

Name          : PowerShell.6
PSVersion     : 6.1
StartupScript :
RunAsUser     :
Permission    : NT AUTHORITY\INTERACTIVE AccessAllowed, BUILTIN\Administrators AccessAllowed, BUILTIN\Remote
                Management Users AccessAllowed

Name          : PowerShell.6.1.0
PSVersion     : 6.1
StartupScript :
RunAsUser     :
Permission    : NT AUTHORITY\INTERACTIVE AccessAllowed, BUILTIN\Administrators AccessAllowed, BUILTIN\Remote
                Management Users AccessAllowed
```

PSDesktop is registered under `Microsoft.PowerShell`, `Microsoft.PowerShell.Workflows`, and `Microsoft.PowerShell32`. The `PowerShell.6` and `PowerShell.6.1.0` are PSCore endpoints, you can have more based on JEA configurations registered on your server.

> My major worry is that most of the manuals I've seen (there were few others) are saying that I need to set basic auth and unnencrypted connections to true, trusted hosts to *

You are right to worry, you should never do this. It means that any traffic is in plaintext and allows anyone to see what you are running, it's output and the credentials you used. The only scenario in which I find this necessary is if you are trying to work on this library, or anything like it, to debug protocol issues. For end users I don't see a reason for this. The `TrustedHosts` option is a client side configuration, you don't need to do anything this this for this library, it's only used when you use PowerShell/WIndows to manage the connection.

> I've read about using self-signed SSL certificates over [HTTPS](https://4sysops.com/archives/powershell-remoting-over-https-with-a-self-signed-ssl-certificate/) but the setup seems to be quite daunting

It can be daunting but I highly recommend it unless you are Kerberos authentication. Technically NTLM does support message encryption but it's quite weak, when you use HTTPS this doesn't matter as encryption is handled by the TLS protocol. The hardest bit for HTTPS is how to generate the self signed certificate, on newer versions of Ansible this easy but older versions it is a bit harder. Ansible has a function you could use that works on all hosts https://github.com/ansible/ansible/blob/devel/examples/scripts/ConfigureRemotingForAnsible.ps1#L89-L163. You can combine that easily to create the https listener with;

```
$subject_name = $env:COMPUTERNAME
$thumbprint = New-LegacySelfSignedCert -SubjectName $subject_name `
    -ValidDays 365

$valueset = @{
    CertificateThumbprint = $thumbprint
    Hostname = $subject_name
}

$selectorset = @{
    Address = "*"
    Transport = "HTTPS"
}

# Delete any existing HTTPS listeners
Remove-WSManInstance -ResourceURI 'winrm/config/Listener' `
    -SelectorSet $selectorset

# Create a new HTTPS listener
New-WSManInstance -ResourceURI 'winrm/config/Listener' `
    -SelectorSet $selectorset `
    -ValueSet $valueset
```

If you have a domain environment you can automatically generate these certs and apply them with GPO and through your AD CS instance as well.

In short, you shouldn't ever use basic auth, use HTTPS if you can but it isn't 100% necessary.

**Comment by danielNemeth19** [comment]
Thank you Jordan for this thorough response. I'll take your comments on board and will try to get HTPPS up and running.
